File: mockserver/mockserver-client-java/src/main/java/org/mockserver/client/A2aMockBuilder.java
package org.mockserver.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.mockserver.mock.Expectation;
import org.mockserver.model.*;

import java.net.URI;
import java.util.*;

import static org.mockserver.model.HttpRequest.request;
import static org.mockserver.model.HttpResponse.response;
import static org.mockserver.model.HttpTemplate.template;
import static org.mockserver.model.HttpTemplate.TemplateType.VELOCITY;
import static org.mockserver.model.JsonRpcBody.jsonRpc;

public class A2aMockBuilder {

    private String path = "/a2a";
    private String agentCardPath = "/.well-known/agent.json";
    private String agentName = "MockAgent";
    private String agentDescription = "A mock A2A agent";
    private String agentVersion = "1.0.0";
    private String agentUrl;
    private final List<A2aSkillDefinition> skills = new ArrayList<>();
    private final List<A2aTaskHandler> taskHandlers = new ArrayList<>();
    private String defaultTaskResponse = "Task completed successfully";
    private boolean streaming = false;
    private String streamingMethod = "message/stream";
    private String pushNotificationUrl;

    private A2aMockBuilder() {
    }

    public static A2aMockBuilder a2aMock() {
        return new A2aMockBuilder();
    }

    public static A2aMockBuilder a2aMock(String path) {
        A2aMockBuilder builder = new A2aMockBuilder();
        builder.path = path;
        return builder;
    }

    public A2aMockBuilder withAgentName(String name) {
        this.agentName = name;
        return this;
    }

    public A2aMockBuilder withAgentDescription(String description) {
        this.agentDescription = description;
        return this;
    }

    public A2aMockBuilder withAgentVersion(String version) {
        this.agentVersion = version;
        return this;
    }

    public A2aMockBuilder withAgentUrl(String url) {
        this.agentUrl = url;
        return this;
    }

    public A2aMockBuilder withAgentCardPath(String path) {
        this.agentCardPath = path;
        return this;
    }

    public A2aMockBuilder withDefaultTaskResponse(String response) {
        this.defaultTaskResponse = response;
        return this;
    }

    /**
     * Advertise and mock the A2A streaming capability. When enabled the agent card reports
     * {@code capabilities.streaming: true} and a streaming JSON-RPC method (default
     * {@code message/stream}, see {@link #withStreamingMethod(String)}) returns an
     * SSE stream of {@code TaskStatusUpdateEvent} and {@code TaskArtifactUpdateEvent}
     * chunks, each wrapped in a JSON-RPC 2.0 response envelope.
     */
    public A2aMockBuilder withStreaming() {
        this.streaming = true;
        return this;
    }

    /**
     * Override the JSON-RPC method that triggers the streaming response. The A2A specification
     * uses {@code message/stream}; the legacy method name is {@code tasks/sendSubscribe}.
     * Implies {@link #withStreaming()}.
     */
    public A2aMockBuilder withStreamingMethod(String method) {
        this.streamingMethod = method;
        this.streaming = true;
        return this;
    }

    /**
     * Advertise and mock A2A push notifications. When configured the agent card reports
     * {@code capabilities.pushNotifications: true}, the {@code tasks/pushNotificationConfig/set}
     * method echoes the registered config, and each {@code tasks/send} additionally POSTs the
     * completed task (the A2A push-notification payload) to the supplied webhook URL while still
     * returning the JSON-RPC task response to the caller.
     *
     * @param webhookUrl absolute webhook URL (e.g. {@code http://localhost:1234/a2a/callback})
     */
    public A2aMockBuilder withPushNotifications(String webhookUrl) {
        this.pushNotificationUrl = webhookUrl;
        return this;
    }

    public A2aSkillBuilder withSkill(String id) {
        return new A2aSkillBuilder(this, id);
    }

    public A2aTaskHandlerBuilder onTaskSend() {
        return new A2aTaskHandlerBuilder(this);
    }

    public Expectation[] applyTo(MockServerClient client) {
        return client.upsert(build());
    }

    public Expectation[] build() {
        List<Expectation> expectations = new ArrayList<>();

        expectations.add(buildAgentCardExpectation());

        for (A2aTaskHandler handler : taskHandlers) {
            expectations.add(buildCustomTaskHandler(handler));
        }

        if (streaming) {
            expectations.add(buildStreamingExpectation());
        }

        if (pushNotificationUrl != null) {
            expectations.add(buildPushNotificationConfigExpectation());
            expectations.add(buildPushNotificationDeliveryExpectation());
        } else {
            expectations.add(buildTasksSendExpectation());
        }
        expectations.add(buildTasksGetExpectation());
        expectations.add(buildTasksCancelExpectation());

        return expectations.toArray(new Expectation[0]);
    }

    private Expectation buildAgentCardExpectation() {
        StringBuilder skillsJson = new StringBuilder("[");
        for (int i = 0; i < skills.size(); i++) {
            if (i > 0) {
                skillsJson.append(", ");
            }
            A2aSkillDefinition skill = skills.get(i);
            skillsJson.append("{");
            skillsJson.append("\"id\": \"").append(escapeJson(skill.id)).append("\"");
            skillsJson.append(", \"name\": \"").append(escapeJson(skill.name != null ? skill.name : skill.id)).append("\"");
            if (skill.description != null) {
                skillsJson.append(", \"description\": \"").append(escapeJson(skill.description)).append("\"");
            }
            if (!skill.tags.isEmpty()) {
                skillsJson.append(", \"tags\": [");
                for (int j = 0; j < skill.tags.size(); j++) {
                    if (j > 0) {
                        skillsJson.append(", ");
                    }
                    skillsJson.append("\"").append(escapeJson(skill.tags.get(j))).append("\"");
                }
                skillsJson.append("]");
            }
            if (!skill.examples.isEmpty()) {
                skillsJson.append(", \"examples\": [");
                for (int j = 0; j < skill.examples.size(); j++) {
                    if (j > 0) {
                        skillsJson.append(", ");
                    }
                    skillsJson.append("\"").append(escapeJson(skill.examples.get(j))).append("\"");
                }
                skillsJson.append("]");
            }
            skillsJson.append("}");
        }
        skillsJson.append("]");

        String url = agentUrl != null ? agentUrl : "http://localhost" + path;

        String agentCardJson = "{" +
            "\"name\": \"" + escapeJson(agentName) + "\", " +
            "\"description\": \"" + escapeJson(agentDescription) + "\", " +
            "\"version\": \"" + escapeJson(agentVersion) + "\", " +
            "\"url\": \"" + escapeJson(url) + "\", " +
            "\"capabilities\": {\"streaming\": " + streaming + ", \"pushNotifications\": " + (pushNotificationUrl != null) + ", \"stateTransitionHistory\": false}, " +
            "\"skills\": " + skillsJson + "}";

        return Expectation.when(
            request().withMethod("GET").withPath(agentCardPath)
        ).thenRespond(
            response().withStatusCode(200)
                .withHeader("Content-Type", "application/json")
                .withBody(agentCardJson)
        );
    }

    private Expectation buildTasksSendExpectation() {
        String resultJson = buildTaskResultJson(defaultTaskResponse, false);
        return Expectation.when(
            request().withMethod("POST").withPath(path).withBody(jsonRpc("tasks/send"))
        ).thenRespond(template(VELOCITY, velocityJsonRpcResponse(resultJson)));
    }

    private Expectation buildTasksGetExpectation() {
        String resultJson = buildTaskResultJson(defaultTaskResponse, false);
        return Expectation.when(
            request().withMethod("POST").withPath(path).withBody(jsonRpc("tasks/get"))
        ).thenRespond(template(VELOCITY, velocityJsonRpcResponse(resultJson)));
    }

    private Expectation buildTasksCancelExpectation() {
        String resultJson = "{\"id\": \"mock-task-id\", \"status\": {\"state\": \"canceled\"}}";
        return Expectation.when(
            request().withMethod("POST").withPath(path).withBody(jsonRpc("tasks/cancel"))
        ).thenRespond(template(VELOCITY, velocityJsonRpcResponse(resultJson)));
    }

    private Expectation buildStreamingExpectation() {
        String text = escapeJson(defaultTaskResponse);
        String taskId = "mock-task-id";

        // A2A streaming: each SSE event data is a JSON-RPC 2.0 response envelope wrapping a
        // TaskStatusUpdateEvent or TaskArtifactUpdateEvent. The JSON-RPC id is not known at
        // build time, so a stable placeholder is used (streaming clients correlate by stream).
        SseEvent statusWorking = SseEvent.sseEvent()
            .withEvent("message")
            .withData("{\"jsonrpc\": \"2.0\", \"id\": \"1\", \"result\": " +
                "{\"taskId\": \"" + taskId + "\", \"kind\": \"status-update\", " +
                "\"status\": {\"state\": \"working\"}, \"final\": false}}");

        SseEvent artifactUpdate = SseEvent.sseEvent()
            .withEvent("message")
            .withData("{\"jsonrpc\": \"2.0\", \"id\": \"1\", \"result\": " +
                "{\"taskId\": \"" + taskId + "\", \"kind\": \"artifact-update\", " +
                "\"artifact\": {\"parts\": [{\"type\": \"text\", \"text\": \"" + text + "\"}]}}}");

        SseEvent statusCompleted = SseEvent.sseEvent()
            .withEvent("message")
            .withData("{\"jsonrpc\": \"2.0\", \"id\": \"1\", \"result\": " +
                "{\"taskId\": \"" + taskId + "\", \"kind\": \"status-update\", " +
                "\"status\": {\"state\": \"completed\"}, \"final\": true}}");

        HttpSseResponse sseResponse = HttpSseResponse.sseResponse()
            .withStatusCode(200)
            .withEvents(statusWorking, artifactUpdate, statusCompleted)
            .withCloseConnection(true);

        Expectation expectation = Expectation.when(
            request().withMethod("POST").withPath(path).withBody(jsonRpc(streamingMethod))
        );
        expectation.thenRespondWithSse(sseResponse);
        return expectation;
    }

    private Expectation buildPushNotificationConfigExpectation() {
        // Echo the registered push-notification config back as the JSON-RPC result.
        String resultJson = "{\"url\": \"" + escapeVelocity(escapeJson(pushNotificationUrl)) + "\"}";
        return Expectation.when(
            request().withMethod("POST").withPath(path).withBody(jsonRpc("tasks/pushNotificationConfig/set"))
        ).thenRespond(template(VELOCITY, velocityJsonRpcResponse(resultJson)));
    }

    private Expectation buildPushNotificationDeliveryExpectation() {
        // When push notifications are configured, a tasks/send both returns the JSON-RPC task
        // response to the caller AND POSTs the completed task (the push-notification payload) to
        // the configured webhook URL. This is modelled with an override-forwarded-request: the
        // request override targets the webhook (literal body), and a Velocity response *template*
        // produces the caller's JSON-RPC response so the request's id is echoed back.
        WebhookTarget target = WebhookTarget.parse(pushNotificationUrl);

        // Literal webhook POST body — no Velocity engine runs over a request override, so only
        // JSON escaping is applied. The push payload carries no JSON-RPC id (server-initiated).
        String pushBody = "{\"jsonrpc\": \"2.0\", \"result\": " + buildTaskResultJsonRaw(defaultTaskResponse, false) + "}";

        HttpRequest webhookRequest = request()
            .withMethod("POST")
            .withPath(target.path)
            .withSocketAddress(target.host, target.port,
                target.secure ? SocketAddress.Scheme.HTTPS : SocketAddress.Scheme.HTTP)
            .withSecure(target.secure)
            .withHeader("Host", target.hostHeader())
            .withHeader("Content-Type", "application/json")
            .withBody(pushBody);

        // Caller response — a Velocity template so $!{request.jsonRpcRawId} echoes the request id,
        // matching the non-push tasks/send contract.
        HttpTemplate clientResponseTemplate = template(VELOCITY,
            velocityJsonRpcResponse(buildTaskResultJson(defaultTaskResponse, false)));

        Expectation expectation = Expectation.when(
            request().withMethod("POST").withPath(path).withBody(jsonRpc("tasks/send"))
        );
        expectation.thenForward(
            HttpOverrideForwardedRequest.forwardOverriddenRequest(webhookRequest)
                .withResponseTemplate(clientResponseTemplate)
        );
        return expectation;
    }

    private Expectation buildCustomTaskHandler(A2aTaskHandler handler) {
        String escapedPattern = handler.messagePattern.replace("/", "\\/");
        escapedPattern = escapedPattern.replace("\n", "\\n").replace("\r", "\\r").replace("\0", "");
        String jsonPathBody = "$[?(@.method == 'tasks/send' && @.params.message.parts[0].text =~ /" + escapedPattern + "/)]";
        String resultJson = buildTaskResultJson(handler.responseText, handler.isError);

        return Expectation.when(
            request().withMethod("POST").withPath(path).withBody(new JsonPathBody(jsonPathBody))
        ).thenRespond(template(VELOCITY, velocityJsonRpcResponse(resultJson)));
    }

    private String buildTaskResultJson(String responseText, boolean isError) {
        // For Velocity-templated response bodies: the text must survive the Velocity engine, so
        // metacharacters are escaped here and un-escaped by the template engine at response time.
        return taskResultJson(escapeVelocity(escapeJson(responseText)), isError);
    }

    private String buildTaskResultJsonRaw(String responseText, boolean isError) {
        // For literal (non-templated) response bodies (e.g. the webhook POST payload), where no
        // Velocity engine runs, only JSON escaping is applied — Velocity escaping would corrupt
        // any '$' / '#' into "${esc.d}" / "${esc.h}".
        return taskResultJson(escapeJson(responseText), isError);
    }

    private String taskResultJson(String escapedText, boolean isError) {
        String state = isError ? "failed" : "completed";
        return "{\"id\": \"mock-task-id\", " +
            "\"status\": {\"state\": \"" + state + "\"}, " +
            "\"artifacts\": [{\"parts\": [{\"type\": \"text\", \"text\": \"" + escapedText + "\"}]}]}";
    }

    private String velocityJsonRpcResponse(String resultJson) {
        return "{\"statusCode\": 200, " +
            "\"headers\": [{\"name\": \"Content-Type\", \"values\": [\"application/json\"]}], " +
            "\"body\": {\"jsonrpc\": \"2.0\", \"result\": " + resultJson + ", \"id\": $!{request.jsonRpcRawId}}}";
    }

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private static String escapeJson(String value) {
        if (value == null) {
            return "";
        }
        try {
            String quoted = OBJECT_MAPPER.writeValueAsString(value);
            return quoted.substring(1, quoted.length() - 1);
        } catch (Exception e) {
            return value.replace("\\", "\\\\").replace("\"", "\\\"")
                .replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t");
        }
    }

    private static String escapeVelocity(String value) {
        if (value == null) {
            return null;
        }
        return value.replace("$", "${esc.d}").replace("#", "${esc.h}");
    }

    static class A2aSkillDefinition {
        final String id;
        String name;
        String description;
        final List<String> tags = new ArrayList<>();
        final List<String> examples = new ArrayList<>();

        A2aSkillDefinition(String id) {
            this.id = id;
        }
    }

    static class A2aTaskHandler {
        final String messagePattern;
        final String responseText;
        final boolean isError;

        A2aTaskHandler(String messagePattern, String responseText, boolean isError) {
            this.messagePattern = messagePattern;
            this.responseText = responseText;
            this.isError = isError;
        }
    }

    static class WebhookTarget {
        final String host;
        final int port;
        final boolean secure;
        final String path;

        private WebhookTarget(String host, int port, boolean secure, String path) {
            this.host = host;
            this.port = port;
            this.secure = secure;
            this.path = path;
        }

        String hostHeader() {
            return host + ":" + port;
        }

        static WebhookTarget parse(String url) {
            URI uri = URI.create(url);
            String scheme = uri.getScheme();
            boolean secure = "https".equalsIgnoreCase(scheme);
            String host = uri.getHost();
            if (host == null) {
                throw new IllegalArgumentException("Invalid push-notification webhook URL (no host): " + url);
            }
            int port = uri.getPort();
            if (port == -1) {
                port = secure ? 443 : 80;
            }
            String path = uri.getRawPath();
            if (path == null || path.isEmpty()) {
                path = "/";
            }
            return new WebhookTarget(host, port, secure, path);
        }
    }

    public static class A2aSkillBuilder {
        private final A2aMockBuilder parent;
        private final A2aSkillDefinition skill;

        A2aSkillBuilder(A2aMockBuilder parent, String id) {
            this.parent = parent;
            this.skill = new A2aSkillDefinition(id);
        }

        public A2aSkillBuilder withName(String name) {
            skill.name = name;
            return this;
        }

        public A2aSkillBuilder withDescription(String description) {
            skill.description = description;
            return this;
        }

        public A2aSkillBuilder withTag(String tag) {
            skill.tags.add(tag);
            return this;
        }

        public A2aSkillBuilder withExample(String example) {
            skill.examples.add(example);
            return this;
        }

        public A2aMockBuilder and() {
            parent.skills.add(skill);
            return parent;
        }
    }

    public static class A2aTaskHandlerBuilder {
        private final A2aMockBuilder parent;
        private String messagePattern = ".*";
        private String responseText = "Task completed";
        private boolean isError = false;

        A2aTaskHandlerBuilder(A2aMockBuilder parent) {
            this.parent = parent;
        }

        public A2aTaskHandlerBuilder matchingMessage(String pattern) {
            this.messagePattern = pattern;
            return this;
        }

        public A2aTaskHandlerBuilder respondingWith(String text) {
            this.responseText = text;
            return this;
        }

        public A2aTaskHandlerBuilder respondingWith(String text, boolean isError) {
            this.responseText = text;
            this.isError = isError;
            return this;
        }

        public A2aMockBuilder and() {
            parent.taskHandlers.add(new A2aTaskHandler(messagePattern, responseText, isError));
            return parent;
        }
    }
}

File: mockserver-client-php/src/A2a/A2aMockBuilder.php
<?php

declare(strict_types=1);

namespace MockServer\A2a;

use MockServer\Expectation;
use MockServer\MockServerClient;

/**
 * Fluent builder for mocking an A2A (Agent-to-Agent) agent speaking JSON-RPC 2.0
 * over HTTP (mirrors {@code org.mockserver.client.A2aMockBuilder}).
 *
 * Produces a list of HTTP expectations whose wire JSON is identical to the Java
 * builder, so every client drives the same server behaviour. The generated
 * control-plane expectations are:
 *
 *  - {@code GET <agentCardPath>} returning the agent card (literal JSON body);
 *  - {@code POST <path>} JSON-RPC {@code tasks/send}, {@code tasks/get},
 *    {@code tasks/cancel} (Velocity-templated JSON-RPC responses);
 *  - per custom task handler, a JSON-PATH-matched {@code tasks/send} responder;
 *  - optionally an SSE streaming responder for {@code message/stream};
 *  - optionally {@code tasks/pushNotificationConfig/set} plus a push-notification
 *    delivery override that POSTs the completed task to the webhook URL while
 *    still returning the JSON-RPC task response to the caller.
 *
 * Note: A2A is purely declarative (matchers + templated/literal responses), so
 * it is fully supported by the REST-only PHP client.
 *
 * @example
 *   A2aMockBuilder::a2aMock('/a2a')
 *       ->withAgentName('WeatherAgent')
 *       ->withSkill('forecast')
 *           ->withName('Forecast')
 *           ->withDescription('Weather forecasting')
 *           ->withTag('weather')
 *       ->and()
 *       ->onTaskSend()
 *           ->matchingMessage('.*weather.*')
 *           ->respondingWith('72F and sunny')
 *       ->and()
 *       ->applyTo($client);
 */
class A2aMockBuilder
{
    private string $path = '/a2a';
    private string $agentCardPath = '/.well-known/agent.json';
    private string $agentName = 'MockAgent';
    private string $agentDescription = 'A mock A2A agent';
    private string $agentVersion = '1.0.0';
    private ?string $agentUrl = null;
    /** @var list<A2aSkillBuilder> */
    private array $skills = [];
    /** @var list<A2aTaskHandlerBuilder> */
    private array $taskHandlers = [];
    private string $defaultTaskResponse = 'Task completed successfully';
    private bool $streaming = false;
    private string $streamingMethod = 'message/stream';
    private ?string $pushNotificationUrl = null;

    public function __construct(string $path = '/a2a')
    {
        $this->path = $path;
    }

    public static function a2aMock(string $path = '/a2a'): self
    {
        return new self($path);
    }

    // --- top-level configuration -------------------------------------------

    public function withAgentName(string $name): self
    {
        $this->agentName = $name;
        return $this;
    }

    public function withAgentDescription(string $description): self
    {
        $this->agentDescription = $description;
        return $this;
    }

    public function withAgentVersion(string $version): self
    {
        $this->agentVersion = $version;
        return $this;
    }

    public function withAgentUrl(string $url): self
    {
        $this->agentUrl = $url;
        return $this;
    }

    public function withAgentCardPath(string $path): self
    {
        $this->agentCardPath = $path;
        return $this;
    }

    public function withDefaultTaskResponse(string $response): self
    {
        $this->defaultTaskResponse = $response;
        return $this;
    }

    /**
     * Advertise and mock the A2A streaming capability. The agent card reports
     * {@code capabilities.streaming: true} and the streaming method (default
     * {@code message/stream}) returns an SSE stream of status/artifact update
     * events wrapped in JSON-RPC 2.0 envelopes.
     */
    public function withStreaming(): self
    {
        $this->streaming = true;
        return $this;
    }

    /**
     * Override the JSON-RPC method that triggers the streaming response and imply
     * {@see withStreaming()}.
     */
    public function withStreamingMethod(string $method): self
    {
        $this->streamingMethod = $method;
        $this->streaming = true;
        return $this;
    }

    /**
     * Advertise and mock A2A push notifications. The agent card reports
     * {@code capabilities.pushNotifications: true}, the
     * {@code tasks/pushNotificationConfig/set} method echoes the registered
     * config, and each {@code tasks/send} additionally POSTs the completed task
     * to {@code $webhookUrl} while still returning the JSON-RPC task response.
     */
    public function withPushNotifications(string $webhookUrl): self
    {
        $this->pushNotificationUrl = $webhookUrl;
        return $this;
    }

    public function withSkill(string $id): A2aSkillBuilder
    {
        return new A2aSkillBuilder($this, $id);
    }

    public function onTaskSend(): A2aTaskHandlerBuilder
    {
        return new A2aTaskHandlerBuilder($this);
    }

    /**
     * @internal Called by {@see A2aSkillBuilder::and()}.
     */
    public function addSkill(A2aSkillBuilder $skill): self
    {
        $this->skills[] = $skill;
        return $this;
    }

    /**
     * @internal Called by {@see A2aTaskHandlerBuilder::and()}.
     */
    public function addTaskHandler(A2aTaskHandlerBuilder $handler): self
    {
        $this->taskHandlers[] = $handler;
        return $this;
    }

    // --- terminal operations -----------------------------------------------

    /**
     * @return list<Expectation>
     */
    public function build(): array
    {
        $expectations = [];

        $expectations[] = $this->buildAgentCardExpectation();

        foreach ($this->taskHandlers as $handler) {
            $expectations[] = $this->buildCustomTaskHandler($handler);
        }

        if ($this->streaming) {
            $expectations[] = $this->buildStreamingExpectation();
        }

        if ($this->pushNotificationUrl !== null) {
            $expectations[] = $this->buildPushNotificationConfigExpectation();
            $expectations[] = $this->buildPushNotificationDeliveryExpectation();
        } else {
            $expectations[] = $this->buildTasksSendExpectation();
        }
        $expectations[] = $this->buildTasksGetExpectation();
        $expectations[] = $this->buildTasksCancelExpectation();

        return $expectations;
    }

    /**
     * @return array<mixed>
     */
    public function applyTo(MockServerClient $client): array
    {
        $results = [];
        foreach ($this->build() as $expectation) {
            $results[] = $client->upsertExpectation($expectation);
        }
        return $results;
    }

    // --- request matchers --------------------------------------------------

    /**
     * @return array<string, mixed>
     */
    private function jsonRpcRequest(string $method): array
    {
        return [
            'method' => 'POST',
            'path' => $this->path,
            'body' => ['type' => 'JSON_RPC', 'method' => $method],
        ];
    }

    /**
     * @param array<string, mixed> $httpRequest
     */
    private function velocityTemplateExpectation(array $httpRequest, string $resultJson): Expectation
    {
        return Expectation::fromArray([
            'httpRequest' => $httpRequest,
            'httpResponseTemplate' => [
                'template' => A2aEscaping::velocityJsonRpcResponse($resultJson),
                'templateType' => 'VELOCITY',
            ],
        ]);
    }

    // --- expectation builders ----------------------------------------------

    private function buildAgentCardExpectation(): Expectation
    {
        $skillItems = [];
        foreach ($this->skills as $skill) {
            $parts = '{';
            $parts .= '"id": "' . A2aEscaping::escapeJson($skill->id) . '"';
            $parts .= ', "name": "' . A2aEscaping::escapeJson($skill->name ?? $skill->id) . '"';
            if ($skill->description !== null) {
                $parts .= ', "description": "' . A2aEscaping::escapeJson($skill->description) . '"';
            }
            if (count($skill->tags) > 0) {
                $tagItems = [];
                foreach ($skill->tags as $tag) {
                    $tagItems[] = '"' . A2aEscaping::escapeJson($tag) . '"';
                }
                $parts .= ', "tags": [' . implode(', ', $tagItems) . ']';
            }
            if (count($skill->examples) > 0) {
                $exampleItems = [];
                foreach ($skill->examples as $example) {
                    $exampleItems[] = '"' . A2aEscaping::escapeJson($example) . '"';
                }
                $parts .= ', "examples": [' . implode(', ', $exampleItems) . ']';
            }
            $parts .= '}';
            $skillItems[] = $parts;
        }
        $skillsJson = '[' . implode(', ', $skillItems) . ']';

        $url = $this->agentUrl ?? 'http://localhost' . $this->path;

        $agentCardJson = '{'
            . '"name": "' . A2aEscaping::escapeJson($this->agentName) . '", '
            . '"description": "' . A2aEscaping::escapeJson($this->agentDescription) . '", '
            . '"version": "' . A2aEscaping::escapeJson($this->agentVersion) . '", '
            . '"url": "' . A2aEscaping::escapeJson($url) . '", '
            . '"capabilities": {"streaming": ' . ($this->streaming ? 'true' : 'false')
            . ', "pushNotifications": ' . ($this->pushNotificationUrl !== null ? 'true' : 'false')
            . ', "stateTransitionHistory": false}, '
            . '"skills": ' . $skillsJson . '}';

        return Expectation::fromArray([
            'httpRequest' => [
                'method' => 'GET',
                'path' => $this->agentCardPath,
            ],
            'httpResponse' => [
                'statusCode' => 200,
                'headers' => [['name' => 'Content-Type', 'values' => ['application/json']]],
                'body' => $agentCardJson,
            ],
        ]);
    }

    private function buildTasksSendExpectation(): Expectation
    {
        return $this->velocityTemplateExpectation(
            $this->jsonRpcRequest('tasks/send'),
            $this->buildTaskResultJson($this->defaultTaskResponse, false)
        );
    }

    private function buildTasksGetExpectation(): Expectation
    {
        return $this->velocityTemplateExpectation(
            $this->jsonRpcRequest('tasks/get'),
            $this->buildTaskResultJson($this->defaultTaskResponse, false)
        );
    }

    private function buildTasksCancelExpectation(): Expectation
    {
        $resultJson = '{"id": "mock-task-id", "status": {"state": "canceled"}}';
        return $this->velocityTemplateExpectation($this->jsonRpcRequest('tasks/cancel'), $resultJson);
    }

    private function buildStreamingExpectation(): Expectation
    {
        $text = A2aEscaping::escapeJson($this->defaultTaskResponse);
        $taskId = 'mock-task-id';

        // A2A streaming: each SSE event data is a JSON-RPC 2.0 response envelope
        // wrapping a TaskStatusUpdateEvent or TaskArtifactUpdateEvent. The
        // JSON-RPC id is not known at build time, so a stable placeholder is used.
        $events = [
            [
                'event' => 'message',
                'data' => '{"jsonrpc": "2.0", "id": "1", "result": '
                    . '{"taskId": "' . $taskId . '", "kind": "status-update", '
                    . '"status": {"state": "working"}, "final": false}}',
            ],
            [
                'event' => 'message',
                'data' => '{"jsonrpc": "2.0", "id": "1", "result": '
                    . '{"taskId": "' . $taskId . '", "kind": "artifact-update", '
                    . '"artifact": {"parts": [{"type": "text", "text": "' . $text . '"}]}}}',
            ],
            [
                'event' => 'message',
                'data' => '{"jsonrpc": "2.0", "id": "1", "result": '
                    . '{"taskId": "' . $taskId . '", "kind": "status-update", '
                    . '"status": {"state": "completed"}, "final": true}}',
            ],
        ];

        return Expectation::fromArray([
            'httpRequest' => $this->jsonRpcRequest($this->streamingMethod),
            'httpSseResponse' => [
                'statusCode' => 200,
                'events' => $events,
                'closeConnection' => true,
            ],
        ]);
    }

    private function buildPushNotificationConfigExpectation(): Expectation
    {
        // Echo the registered push-notification config back as the JSON-RPC result.
        $url = A2aEscaping::escapeVelocity(A2aEscaping::escapeJson($this->pushNotificationUrl));
        $resultJson = '{"url": "' . $url . '"}';
        return $this->velocityTemplateExpectation(
            $this->jsonRpcRequest('tasks/pushNotificationConfig/set'),
            $resultJson
        );
    }

    private function buildPushNotificationDeliveryExpectation(): Expectation
    {
        // When push notifications are configured, a tasks/send both returns the
        // JSON-RPC task response to the caller AND POSTs the completed task to the
        // configured webhook URL. This is modelled with an
        // httpOverrideForwardedRequest: the request override targets the webhook
        // (literal body), and a Velocity response *template* produces the caller's
        // JSON-RPC response so the request's id is echoed back.
        $target = A2aEscaping::parseWebhook((string) $this->pushNotificationUrl);

        // Literal webhook POST body — no Velocity engine runs over a request
        // override, so only JSON escaping is applied. The push payload carries no
        // JSON-RPC id (server-initiated).
        $pushBody = '{"jsonrpc": "2.0", "result": '
            . $this->buildTaskResultJsonRaw($this->defaultTaskResponse, false) . '}';

        $hostHeader = $target['host'] . ':' . $target['port'];

        $webhookRequest = [
            'method' => 'POST',
            'path' => $target['path'],
            'socketAddress' => [
                'host' => $target['host'],
                'port' => $target['port'],
                'scheme' => $target['secure'] ? 'HTTPS' : 'HTTP',
            ],
            'secure' => $target['secure'],
            'headers' => [
                'Host' => [$hostHeader],
                'Content-Type' => ['application/json'],
            ],
            'body' => $pushBody,
        ];

        // Caller response — a Velocity template so $!{request.jsonRpcRawId} echoes
        // the request id, matching the non-push tasks/send contract.
        $clientResponseTemplate = [
            'template' => A2aEscaping::velocityJsonRpcResponse(
                $this->buildTaskResultJson($this->defaultTaskResponse, false)
            ),
            'templateType' => 'VELOCITY',
        ];

        return Expectation::fromArray([
            'httpRequest' => $this->jsonRpcRequest('tasks/send'),
            'httpOverrideForwardedRequest' => [
                'requestOverride' => $webhookRequest,
                'responseTemplate' => $clientResponseTemplate,
            ],
        ]);
    }

    private function buildCustomTaskHandler(A2aTaskHandlerBuilder $handler): Expectation
    {
        $escapedPattern = $this->escapeMessagePattern($handler->messagePattern);
        $jsonPathBody = "$[?(@.method == 'tasks/send' && @.params.message.parts[0].text =~ /"
            . $escapedPattern . "/)]";
        $resultJson = $this->buildTaskResultJson($handler->responseText, $handler->isError);

        return $this->velocityTemplateExpectation(
            [
                'method' => 'POST',
                'path' => $this->path,
                'body' => ['type' => 'JSON_PATH', 'jsonPath' => $jsonPathBody],
            ],
            $resultJson
        );
    }

    /**
     * Escape a user-supplied regular expression so it can be embedded between the
     * '/' delimiters of the JsonPath {@code =~ /<PATTERN>/} matcher without the
     * pattern being able to break OUT of the regex literal.
     *
     * Single-pass, left-to-right, byte-oriented (all special chars are ASCII;
     * other bytes are emitted verbatim to preserve UTF-8):
     *  - On '\\': preserve an existing escape sequence by emitting the backslash
     *    plus the following char verbatim (so '\d' stays '\d', '\/' stays '\/');
     *    a TRAILING lone backslash is doubled to a literal backslash so it cannot
     *    escape the closing delimiter.
     *  - On bare '/': emit '\/' so it cannot terminate the regex literal.
     *  - On newline / CR: emit the two-char escapes '\n' / '\r'.
     *  - On NUL: strip it.
     *  - Otherwise: emit the byte verbatim.
     */
    private function escapeMessagePattern(string $pattern): string
    {
        $out = '';
        $len = strlen($pattern);
        for ($i = 0; $i < $len; $i++) {
            $c = $pattern[$i];
            if ($c === '\\') {
                if ($i + 1 < $len) {
                    $out .= '\\' . $pattern[$i + 1];
                    $i++;
                } else {
                    // trailing lone backslash -> literal backslash
                    $out .= '\\\\';
                }
            } elseif ($c === '/') {
                $out .= '\\/';
            } elseif ($c === "\n") {
                $out .= '\\n';
            } elseif ($c === "\r") {
                $out .= '\\r';
            } elseif ($c === "\0") {
                // strip NUL
                continue;
            } else {
                $out .= $c;
            }
        }
        return $out;
    }

    /**
     * For Velocity-templated response bodies: the text must survive the Velocity
     * engine, so metacharacters are escaped here and un-escaped at response time.
     */
    private function buildTaskResultJson(string $responseText, bool $isError): string
    {
        return $this->taskResultJson(
            (string) A2aEscaping::escapeVelocity(A2aEscaping::escapeJson($responseText)),
            $isError
        );
    }

    /**
     * For literal (non-templated) response bodies (e.g. the webhook POST payload),
     * where no Velocity engine runs, only JSON escaping is applied — Velocity
     * escaping would corrupt any '$' / '#' into "${esc.d}" / "${esc.h}".
     */
    private function buildTaskResultJsonRaw(string $responseText, bool $isError): string
    {
        return $this->taskResultJson(A2aEscaping::escapeJson($responseText), $isError);
    }

    private function taskResultJson(string $escapedText, bool $isError): string
    {
        $state = $isError ? 'failed' : 'completed';
        return '{"id": "mock-task-id", '
            . '"status": {"state": "' . $state . '"}, '
            . '"artifacts": [{"parts": [{"type": "text", "text": "' . $escapedText . '"}]}]}';
    }
}
